README


METRIC DAEMON


Daemon the generates example metric values and makes them available via
a UDP channel.


Platforms

Should work on all POSIX platforms including windows using cygwin


Prerequistes

-   gcc (Compile package)
-   wget (Download package)


Installation

1.  Download the current release:

        wget https://github.com/BMCTrueSight/metric-daemon/archive/v0.0.1.tar.gz

2.  Extract the contents: ```bash tar xvf v0.0.1.tar.gz

3.  Change directory to distribution

        cd blah

4.  Configure:

        $ ./configure

5.  Compile the distribution

        $ make

6.  Install the binaries

        $ sudo make install


Operation

You can start the daemon in the foreground by running the following:

    ```
    $ metricd
    The server UDP port number is 12345
    ```

You can test that the daemon is working correctly by using the
accompanying client command:

    ```
    $ metric-client 127.0.0.1 12345 bytecount
    request: BYTECOUNT, response: 95514.000
    ```


TODO

-   Script to run as true daemon via init.d or systemd.



NEWS
2016-06-15 Initial release of code


src/main.c
//
// Copyright 2016 BMC Software, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <netdb.h>

#define MAXBUF 1024*1024

const char *m_count = "COUNT";
const char *m_status = "STATUS";

double process_request(const char *request) {
    double value = 0;
    if (strcmp(request, m_count) == 0) {
        value = 1.0;
    } else if (strcmp(request, m_status) == 0) {
        value = 2.0;
    } else {
        value = -1.0;
    }
    return value;
}

void serve_measurements(int sd) {
    socklen_t len;
    int n;
    char bufin[MAXBUF];
    struct sockaddr_in remote;
    char request[MAXBUF + 1];

    // need to know how big address struct is, len must be set before the call to recvfrom()
    len = sizeof(remote);

    while (1) {
        // read a datagram from the socket (put result in bufin) */
        n = recvfrom(sd, bufin, MAXBUF, 0, (struct sockaddr *) &remote, &len);

        if (n < 0) {
            perror("Error receiving data");
        } else {
            strncpy(request, bufin, n);
            request[n - 1] = '\0';

            // Output the address of sender and their metric request
            fprintf(stderr, "Request for metric %s from %s port %d\n", request,
                    inet_ntoa(remote.sin_addr), ntohs(remote.sin_port));

            double measurement = process_request(request);
            n = sprintf(bufin, "%f", measurement);
            fprintf(stderr, "%s\n", bufin);
            /* Got something, just send it back */
            sendto(sd, bufin, n, 0, (struct sockaddr *) &remote, len);
        }
    }
}

int main(int argc, char *argv[]) {
    int ld;
    struct sockaddr_in skaddr;
    socklen_t length;

    // create a socket
    // IP protocol family (PF_INET)
    // UDP protocol (SOCK_DGRAM)
    if ((ld = socket(PF_INET, SOCK_DGRAM, 0)) < 0) {
        fprintf(stderr, "Problem creating socket\n");
        exit(1);
    }

    // establish our address
    // address family is AF_INET
    // our IP address is INADDR_ANY (any of our IP addresses)
    // the port number is assigned by the kernel
    skaddr.sin_family = AF_INET;
    skaddr.sin_addr.s_addr = htonl(INADDR_ANY);
    skaddr.sin_port = htons(0);

    if (bind(ld, (struct sockaddr *) &skaddr, sizeof(skaddr)) < 0) {
        fprintf(stderr, "Problem binding\n");
        exit(0);
    }

    // find out what port we were assigned and print it out

    length = sizeof(skaddr);
    if (getsockname(ld, (struct sockaddr *) &skaddr, &length) < 0) {
        fprintf(stderr, "Error getsockname\n");
        exit(1);
    }

    // port number's are network byte order, we have to convert to
    // host byte order before printing !
    fprintf(stderr, "The server UDP port number is %d\n", ntohs(skaddr.sin_port));

    // Handle requests for measurements
    serve_measurements(ld);
    return 0;
}

src/client.c
//
// Copyright 2016 BMC Software, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
#include <ctype.h>
#include <libgen.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <netdb.h> 

#define BUFSIZE 1024

char program_name[BUFSIZE];

/* 
 * error - wrapper for perror
 */
void error(char *msg) {
    perror(msg);
    exit(0);
}

void usage() {
    fprintf(stderr, "usage: %s <hostname> <port> <bytecount|duration|number|percent>\n",
            basename(program_name));
    exit(0);
}

char *upper(char *s) {
    for (int i = 0; s[i]; i++) {
        s[i] = toupper(s[i]);
    }
    return s;
}

static const char *requests[] = {
        "bytecount",
        "duration",
        "number",
        "percent",
        NULL
};

int validate_request(const char *request) {
    int result = 1;

    for (int i = 0; requests[i]; i++) {
        if (strcmp(requests[i], request) == 0) {
            result = 0;
            break;
        }
    }
    return result;
}


int main(int argc, char **argv) {
    int sockfd, portno, n;
    socklen_t serverlen;
    struct sockaddr_in serveraddr;
    struct hostent *server;
    char *hostname;
    char request[BUFSIZE];
    char response[BUFSIZE];

    // Set our request/response buffers to empty values
    memset(request, '\0', BUFSIZE);
    memset(response, '\0', BUFSIZE);

    strcpy(program_name, argv[0]);

    // check command line arguments
    if (argc < 4) {
        usage();
    }

    hostname = argv[1];
    portno = atoi(argv[2]);
    if (validate_request(argv[3])) {
        usage();
    }
    else {
        memcpy(request, upper(argv[3]), strlen(argv[3]));
    }

    /* socket: create the socket */
    sockfd = socket(AF_INET, SOCK_DGRAM, 0);
    if (sockfd < 0)
        error("ERROR opening socket");

    /* gethostbyname: get the server's DNS entry */
    server = gethostbyname(hostname);
    if (server == NULL) {
        fprintf(stderr, "ERROR, no such host as %s\n", hostname);
        exit(0);
    }

    /* build the server's Internet address */
    memset(&serveraddr, '\0', sizeof(serveraddr));
    serveraddr.sin_family = AF_INET;
    memcpy(server->h_addr, &serveraddr.sin_addr.s_addr, server->h_length);
    serveraddr.sin_port = htons(portno);

    /* send the message to the server */
    serverlen = sizeof(serveraddr);
    n = sendto(sockfd, request, strlen(request), 0, (struct sockaddr *) &serveraddr, serverlen);
    if (n < 0) {
        error("ERROR in sendto");
    }

    // Extract the servers response
    n = recvfrom(sockfd, response, BUFSIZE, 0, (struct sockaddr *) &serveraddr, &serverlen);
    if (n < 0) {
        error("ERROR in recvfrom");
    }
    fprintf(stdout, "request: \"%s\", response: %s\n", request, response);
    fflush(stdout);
    return 0;
}


src/server.c
//
// Copyright 2016 BMC Software, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
#include <math.h>
#include <time.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <netdb.h>

#define MAXBUF 1024*1024

const char *m_bytecount = "BYTECOUNT";
const char *m_duration = "DURATION";
const char *m_number = "NUMBER";
const char *m_percent = "PERCENT";

double rand_range(int min, int max) {
    double rnum =rand()/(double)RAND_MAX;
    return floor(rnum * ((max - min) + 1)) + min;
}

int process_request(const char *request, double *value) {
    int result = 1;
    if (strcmp(request, m_bytecount) == 0) {
        *value = rand_range(123, 123456);
    } else if (strcmp(request, m_duration) == 0) {
        *value = rand_range(0, 10000);
    } else if (strcmp(request, m_number) == 0) {
        *value = rand_range(0, 100);
    } else if (strcmp(request, m_percent) == 0) {
        *value = rand_range(0,100)/100.0;
    } else {
        result = 0;
    }
    return result;
}

void serve_measurements(int sd) {
    socklen_t len;
    int n;
    char bufin[MAXBUF];
    struct sockaddr_in remote;
    char request[MAXBUF + 1];

    // need to know how big address struct is, len must be set before the call to recvfrom()
    len = sizeof(remote);

    while (1) {
        // read a datagram from the socket (put result in bufin) */
        n = recvfrom(sd, bufin, MAXBUF, 0, (struct sockaddr *) &remote, &len);

        if (n < 0) {
            perror("Error receiving data");
        } else {
            strncpy(request, bufin, n);
            request[n - 1] = '\0';

            double measurement = 0;
	    int n = 0;
		    
            if (process_request(request, &measurement)) {
                // Output the address of sender and their metric request
                fprintf(stderr, "from: %s port: %d, request: %s, response: %.3f\n",
                        inet_ntoa(remote.sin_addr), ntohs(remote.sin_port),
	                request, measurement);
                n = sprintf(bufin, "%.3f", measurement);
	    } else {
            	n = sprintf(bufin, "%s", "UNKNOWN_METRIC");
	    }
            /* Got something, just send it back */
            sendto(sd, bufin, n, 0, (struct sockaddr *) &remote, len);
        }
    }
}

int main(int argc, char *argv[]) {
    int ld;
    struct sockaddr_in skaddr;
    socklen_t length;
    int port = 12345;

    if (argc == 2) {
        port = atoi(argv[2]);
    }

    srand(time(NULL));

    // create a socket
    // IP protocol family (PF_INET)
    // UDP protocol (SOCK_DGRAM)
    if ((ld = socket(PF_INET, SOCK_DGRAM, 0)) < 0) {
        fprintf(stderr, "Problem creating socket\n");
        exit(1);
    }

    // establish our address
    // address family is AF_INET
    // our IP address is INADDR_ANY (any of our IP addresses)
    // the port number is assigned by the kernel
    skaddr.sin_family = AF_INET;
    skaddr.sin_addr.s_addr = htonl(INADDR_ANY);
    skaddr.sin_port = htons(port);

    if (bind(ld, (struct sockaddr *) &skaddr, sizeof(skaddr)) < 0) {
        fprintf(stderr, "Problem binding\n");
        exit(0);
    }

    // find out what port we were assigned and print it out

    length = sizeof(skaddr);
    if (getsockname(ld, (struct sockaddr *) &skaddr, &length) < 0) {
        fprintf(stderr, "Error getsockname\n");
        exit(1);
    }

    // port number's are network byte order, we have to convert to
    // host byte order before printing !
    fprintf(stderr, "The server UDP port number is %d\n", ntohs(skaddr.sin_port));

    // Handle requests for measurements
    serve_measurements(ld);
    return 0;
}
